### PR TITLE
Initial version of the private references check.

### PR DESCRIPTION
**Pull request by VelinYordanov**
Checks if a bundle uses internal types in its public API.
See #210

Signed-off-by: VelinYordanov <velin.iordanov@gmail.com>

### CODE REVIEW

**Comment by martinvw** [comment]
To be able to merge this I think  we should first reach a conclusion regarding: https://github.com/eclipse/smarthome/issues/4330

**Comment by maggu2810** [comment]
Why? Hasn't it been about checks so, no internal stuff is used in the public API of exported stuff?
This should be independent of the respective issue at all. But something that should be checked for correct modularization at all.

The conclusion of the respective issue should be only about if handlers should be part of an exported package but should be independent of a private reference check.

I forgot about that issue, but for me it is okay to export handlers if there are use cases for it.

**Comment by martinvw** [comment]
But to me what you mention sounds very related to  

> Checks if a bundle uses internal types in its public API.

Because openHAB bundles mainly fail to comply because the handlers are exported by default and in most implementations we have to drag everything to the public api or rewrite the complete binding, not exporting the handlers makes more sense to me for most bindings.

The reason I was triggered is because there are some openHAB PR which drag a lot of code to public API to resolve this issue.

**Comment by martinvw** [comment]
I just now see it is a warning and not an error so that makes the order less important if I remember correctly

File: custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/PrivateReferencesCheck.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    private void collectImportsForFile(Collection<DetailAST> differentNodeTypes, String fullTypeName) {
+        Collection<String> imports = differentNodeTypes.stream()
+                .filter(x -> x.getType() == TokenTypes.IMPORT)
+                .map(SatCheckUtils::createFullyQualifiedName)
+                .collect(Collectors.toList());
+
+        fullTypeNamesToImports.put(fullTypeName, imports);
+    }
+
+    private String getFullTypeName(FileText fileText) {
```
[the remark this answers is not in the document]
Comment (VelinYordanov) [reply custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/PrivateReferencesCheck.java]: The methods require AST - in this check if the type is internal we only need the package declaration. I personally don't think we should create an AST only for the package declaration.

**Comment by VelinYordanov** [comment]
Seems like eclipse/smarthome#4330 is closed @martinvw how should we proceed?

**Comment by martinvw** [comment]
Imho, we could make this a warning or maybe an info and then slowly expose less, because I don't think exposing more is the answer. We can also see that in the solution for eclipse/smarthome#4330 there it is also resolved by moving more things internally.

**Comment by dstivanov** [comment]
@VelinYordanov : Are you satisfied with the answer or need more clarifications ?

**Comment by kaikreuzer** [comment]
Closing due to inactivity.

**Pull request closed without merging**